### PR TITLE
List all rewards per user

### LINKED ISSUE #47: List of pending rewards per user in external staking
Bonus: it would be nice to have a list of all pending rewards for a user.

I guess you could use `self.stakes.prefix(&user).range(...)` to get that list and return caluclate_reward for each one.

Even without pagination, it would be very useful for any UI

_Originally posted by @ethanfrey in https://github.com/osmosis-labs/mesh-security/pull/46#discussion_r1217896481_

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #47 

Helpful query to view all rewards a user has over all validators when cross-staking

### CODE REVIEW

File: contracts/provider/external-staking/src/msg.rs [hunk 1]
```diff
@@ -89,12 +89,33 @@ pub struct UsersResponse {
...
 }
 
-/// Response for penging rewards query
+/// Response for pending rewards query on one validator
 #[cw_serde]
 pub struct PendingRewards {
     pub amount: Coin,
 }
 
+/// Response for pending rewards query on one validator
```
Comment (maurolacy) [inline contracts/provider/external-staking/src/msg.rs:98]:
```suggestion
/// Response for pending rewards query on all validators
```

File: contracts/provider/external-staking/src/msg.rs [hunk 2]
```diff
@@ -89,12 +89,33 @@ pub struct UsersResponse {
...
+#[cw_serde]
+pub struct AllPendingRewards {
+    pub rewards: Vec<ValidatorPendingReward>,
+}
+
+#[cw_serde]
+pub struct ValidatorPendingReward {
+    pub amount: Coin,
+    pub validator: String,
```
Comment (maurolacy) [inline contracts/provider/external-staking/src/msg.rs:107]: Nit: Switch the order of these for clarity.

File: contracts/provider/external-staking/src/msg.rs [hunk 3]
```diff
@@ -89,12 +89,33 @@ pub struct UsersResponse {
...
+}
+
+impl ValidatorPendingReward {
+    pub fn new(amount: u128, denom: impl Into<String>, validator: impl Into<String>) -> Self {
```
Comment (maurolacy) [inline contracts/provider/external-staking/src/msg.rs:111]: Here too, I would put `validator` first, and then `amount` and `denom`.

File: contracts/provider/external-staking/src/contract.rs [hunk 4]
```diff
@@ -793,17 +793,54 @@ impl ExternalStakingContract<'_> {
...
+            .take(limit)
+            .map(|item| {
+                let (validator, stake_lock) = item?;
+                let stake = stake_lock.read()?;
+                let distribution = self
+                    .distribution
+                    .may_load(ctx.deps.storage, &validator)?
+                    .unwrap_or_default();
+                let amount = Self::calculate_reward(stake, &distribution)?;
+                Ok::<_, ContractError>(ValidatorPendingReward {
```
Comment (maurolacy) [inline contracts/provider/external-staking/src/contract.rs:826]: Nit: use the `new` helper.

File: contracts/provider/external-staking/src/contract.rs [hunk 5]
```diff
@@ -793,17 +793,54 @@ impl ExternalStakingContract<'_> {
...
+        let config = self.config.load(ctx.deps.storage)?;
+
+        let rewards: Vec<_> = self
+            .stakes
+            .prefix(&user)
+            .range(ctx.deps.storage, bound, None, Order::Ascending)
+            .take(limit)
+            .map(|item| {
+                let (validator, stake_lock) = item?;
+                let stake = stake_lock.read()?;
```
Comment (maurolacy) [inline contracts/provider/external-staking/src/contract.rs:820]: Use a `MaybeReward` wrapper not to error here. I can do it in a follow-up PR.

**Review by maurolacy — APPROVED** [review]
LGTM.

File: contracts/provider/external-staking/src/contract.rs (hunk 5 above)
Comment (ethanfrey) [reply contracts/provider/external-staking/src/contract.rs:820]: Yeah, seems like nice follow up. You should also update pending_rewards to do the same (that errors on lock as well).
I just made these two work the same, happy for follow up to improve them both (and a test for querying with in-flight tx)

Comment (maurolacy) [reply contracts/provider/external-staking/src/contract.rs:820]: See #86.

**Pull request merged**